src/common/tests_dmg_helpers.c

```diff
@@ -578,3 +576,158 @@ dmg_pool_list(const char *dmg_config_file, const char *group,
 
 	return rc;
 }
+
+static int
+parse_device_info(struct json_object *smd_dev, device_list *devices,
+		  char *host, int dev_length, int *disks)
+{
+	struct json_object	*tmp;
+	struct json_object	*dev = NULL;
+	int			i;
+
+	for (i = 0; i < dev_length; i++) {
+		dev = json_object_array_get_idx(smd_dev, i);
+		strcpy(devices[*disks].host, strtok(host, ":") + 1);
+		if (!json_object_object_get_ex(dev, "uuid", &tmp)) {
+			D_ERROR("unable to extract uuid from JSON\n");
+		return -DER_INVAL;
+		}
+		uuid_parse(json_object_get_string(tmp),
+			   devices[*disks].device_id);
+
+		if (!json_object_object_get_ex(dev, "state", &tmp)) {
+			D_ERROR("unable to extract state from JSON\n");
+			return -DER_INVAL;
+		}
+		strcpy(devices[*disks].state, json_object_to_json_string(tmp));
+
+		if (!json_object_object_get_ex(dev, "rank", &tmp)) {
+			D_ERROR("unable to extract rank from JSON\n");
+		return -DER_INVAL;
+		}
+		devices[*disks].rank = atoi(json_object_to_json_string(tmp));
+		*disks = *disks + 1;
+	}
+
+	return 0;
+}
+
+int
+dmg_storage_device_list(const char *dmg_config_file, int *ndisks,
+			device_list *devices)
+{
+	struct json_object	*dmg_out = NULL;
+	struct json_object	*storage_map = NULL;
+	struct json_object	*hosts = NULL;
+	struct json_object	*smd_info = NULL;
+	struct json_object	*smd_dev = NULL;
+	char		host[100];
+	int			dev_length = 0;
+	int			rc = 0;
+	int			*disk;
+
+	if (ndisks != NULL)
+		*ndisks = 0;
+
+	disk = (int *)calloc(1, sizeof(int));
+	rc = daos_dmg_json_pipe("storage query list-devices", dmg_config_file,
+				NULL, 0, &dmg_out);
+	if (rc != 0) {
+		D_ERROR("dmg failed");
+		goto out_json;
+	}
+
+	if (!json_object_object_get_ex(dmg_out, "host_storage_map",
+				       &storage_map)) {
+		D_ERROR("unable to extract host_storage_map from JSON\n");
+		return -DER_INVAL;
+	}
+
+	json_object_object_foreach(storage_map, key, val) {
+		D_DEBUG(DB_TEST, "key:\"%s\",val=%s\n", key,
+			json_object_to_json_string(val));
+
+		if (!json_object_object_get_ex(val, "hosts", &hosts)) {
+			D_ERROR("unable to extract hosts from JSON\n");
+			return -DER_INVAL;
+		}
+
+		strcpy(host, json_object_to_json_string(hosts));
+
+		json_object_object_foreach(val, key1, val1) {
+			D_DEBUG(DB_TEST, "key1:\"%s\",val1=%s\n", key1,
+				json_object_to_json_string(val1));
+
+			json_object_object_get_ex(val1, "smd_info", &smd_info);
+			if (smd_info != NULL) {
+				if (!json_object_object_get_ex(
+					smd_info, "devices", &smd_dev)) {
+					D_ERROR("unable to extract devices\n");
+					return -DER_INVAL;
+				}
+
+				if (smd_dev != NULL)
+					dev_length = json_object_array_length(
+						smd_dev);
+
+				if (ndisks != NULL)
+					*ndisks = *ndisks + dev_length;
+
+				if (devices != NULL) {
+					rc = parse_device_info(smd_dev, devices,
+							       host, dev_length,
+							       disk);
+					if (rc != 0)
+						goto out_json;
+				}
+			}
+		}
+	}
+
+out_json:
+	if (dmg_out != NULL)
+		json_object_put(dmg_out);
+
+	D_FREE(disk);
```

Comment:
Mixing calloc with D_FREE will break the cause a false leak in log parsing tool

Reply:
Jeff is correct, and we should use the macros but I don't think the unit tests are using the log parsing tool.